### PR TITLE
BigT,Map, Stream Changes for Minibase

### PR DESCRIPTION
**Pull request by nagarjun247**

### CODE REVIEW

File: src/BigT/bigt.java [hunk 1]
```diff
@@ -0,0 +1,564 @@
...
+    public BTreeFile utilityIndex = null;
+    public String indexUtil;
+    private AttrType[] attrType;
+    private FldSpec[] projlist;
+    private CondExpr[] expr;
+    MapIndexScan index_scan;
+    private int insertType;
+    short[] res_str_sizes = new short[]{Map.DEFAULT_ROW_LABEL_ATTRIBUTE_SIZE, Map.DEFAULT_STRING_ATTRIBUTE_SIZE, Map.DEFAULT_STRING_ATTRIBUTE_SIZE};
+
+    public bigt(String name, boolean insert) throws HFException, HFBufMgrException, HFDiskMgrException, IOException,
```
Comment (lowkya) [inline src/BigT/bigt.java]: According to the Phase II doc :

bigt(java.lang.String name, int type) -  Initialize the big table. type is an integer be- tween 1 and 5 and the different types will correspond to different clustering and indexing strategies you will use for the bigtable.

Your second param is a boolean.

File: src/BigT/bigt.java [hunk 2]
```diff
@@ -0,0 +1,565 @@
...
+                    deleteAllNodesInIndex(indexFiles.get(i));
+                }
+            } catch(Exception e) {
+                e.printStackTrace();
+                System.out.println("Exception occurred while destroying all nodes of index files");
+            }
+        }
+        initExprs();
+    }
+
```
Comment (lowkya) [inline src/BigT/bigt.java:72]: I think you are missing this method. void deleteBigt(): Delete the bigtable from the database.

File: src/diskmgr/PCounter.java [hunk 3]
```diff
@@ -0,0 +1,31 @@
+package diskmgr;
+
+public class PCounter {
+    public static int rcounter;
+    public static int wcounter;
+
+    public PCounter() {
```
Comment (lowkya) [inline src/diskmgr/PCounter.java]: Can you delete this file ? I have made changes for this including changes to DB.java and raised a PR.
https://github.com/nagarjun247/javaminibase/pull/1
Have a look at it. 
Thanks.

File: src/BigT/Map.java [hunk 4]
```diff
@@ -0,0 +1,277 @@
...
+        fldCnt = fromMap.noOfFlds();
+        fldOffset = fromMap.copyFldOffset();
+    }
+
+    public Map(int size){
+        data = new byte[size];
+        map_offset = 0;
+        map_length = size;
+    }
+
```
Comment (lowkya) [inline src/BigT/Map.java:62]: You are missing this.
Map(byte[] amap, int offset): Construct a map from a byte array.

File: src/BigT/Stream.java [hunk 5]
```diff
@@ -0,0 +1,303 @@
...
+    MapIterator map_iterator;
+    SortMap sortMap;
+
+    CondExpr[] cond_exprs;
+    CondExpr[] index_2_filter;
+    CondExpr[] index_3_filter;
+    CondExpr[] index_4_filter;
+    CondExpr[] index_5_filter;
+
+    public Stream(String bigtName, int indexType, int orderType, String rowFilter, String columnFilter, String valueFilter, int numBuf) {
```
Comment (lowkya) [inline src/BigT/Stream.java]: Stream(bigt bigtable, int orderType, java.lang.String rowFilter, java.lang.String columnFilter, java.lang.String valueFilter): Initialize a stream of maps on bigtable.
This is how it is in the phase II doc.

File: src/BigT/Stream.java [hunk 6]
```diff
@@ -0,0 +1,303 @@
...
+            }else{
+                expr.operand2.string = filter;
+            }
+            expr.next = null;
+            result.add(expr);
+        }
+        return result;
+    }
+
+    public Map getNext() {
```
Comment (lowkya) [inline src/BigT/Stream.java:277]: Map getNext(MID mid): Retrieve the next map in the stream.

File: src/heap/Heapfile.java [hunk 7]
```diff
@@ -599,6 +601,61 @@ public RID insertRecord(byte[] recPtr)
       return rid;
       
     }
+
```
Comment (lowkya) [inline src/heap/Heapfile.java:447]: Some of the below are missing.

– Heapfile(java.lang.String name): Initialize.
– void deleteFile(): Delete the file from the database.
– boolean deleteMap(MID mid): Delete map from file with given mid.
– int getMapCnt(): Return number of maps in the file.
– Map getMap(MID mid): Read the map from file.
– MID insertMap(byte[] mapPtr) Insert map into file, return its Mid.
– Scan openScan(): Initiate a sequential scan.
– boolean updateMap(MID mid, Map newmap): Updates the specified map in the heapfile.

File: src/BigT/bigt.java [hunk 8]
```diff
@@ -0,0 +1,565 @@
...
+            System.err.println("Error in creating utility index");
+            ex.printStackTrace();
+        }
+    }
+
+    public void insertIndex(MID mid, Map map, int type) throws KeyTooLongException, KeyNotMatchException, LeafInsertRecException,
+            IndexInsertRecException, ConstructPageException, UnpinPageException, PinPageException,
+            NodeNotMatchException, ConvertException, DeleteRecException, IndexSearchException, IteratorException,
+            LeafDeleteException, InsertException, IOException {
+        switch (type) {
```
Comment (lowkya) [inline src/BigT/bigt.java:182]: This requires a change in BTTreeFile too. Do check.

File: src/BigT/bigt.java [hunk 9]
```diff
@@ -0,0 +1,565 @@
...
+    public boolean removeIndex(MID mid, Map map, int type)
+            throws KeyTooLongException, KeyNotMatchException, LeafInsertRecException, IndexInsertRecException,
+            ConstructPageException, UnpinPageException, PinPageException, NodeNotMatchException, ConvertException,
+            DeleteRecException, IndexSearchException, IteratorException, LeafDeleteException, InsertException,
+            IOException, DeleteFashionException, LeafRedistributeException, RedistributeException, InsertRecException,
+            FreePageException, RecordNotFoundException, IndexFullDeleteException {
+        boolean status;
+        if(type == 1){
+            status = true;
+        }else if(type == 2){
```
Comment (lowkya) [inline src/BigT/bigt.java:209]: This requires a change in BTTreeFile too. Do check.

File: src/BigT/bigt.java [hunk 10]
```diff
@@ -0,0 +1,565 @@
...
+        return getCount(3);
+    }
+
+    public int getColumnCnt()  throws Exception{
+        return getCount(4);
+    }
+
+    public int getCount(int orderType) throws Exception{
+        int numBuf = (int)((SystemDefs.JavabaseBM.getNumBuffers()*3)/4);
+        Stream stream = new Stream(this.name, null, 1,  orderType, "*", "*", "*", numBuf);
```
Comment (lowkya) [inline src/BigT/bigt.java]: The params to the constructer are wrong. There is a compile-time error. Please check.

File: src/heap/Heapfile.java [hunk 11]
```diff
@@ -1,9 +1,11 @@
 package heap;
 
 import java.io.*;
```
Comment (lowkya) [inline src/heap/Heapfile.java:3]: A few rid references still exist in this. There are shown as compile-time errors.
Ex: for (mid = currentDirPage.firstRecord(); - Line 633

File: src/diskmgr/PCounter.java (hunk 3 above)
Comment (nagarjun247) [reply src/diskmgr/PCounter.java]: Done

File: src/BigT/bigt.java [hunk 12]
```diff
@@ -0,0 +1,509 @@
...
+                }
+            }
+            t = stream.getNext();
+        }
+        stream.closestream();
+        return count;
+    }
+
+
+    public MID insertMap(Map map, int type) throws HFDiskMgrException,
```
Comment (lowkya) [inline src/BigT/bigt.java:272]: MID insertMap(byte[] mapPtr) Insert map into the big table, return its Mid. The insertMap() method ensures that there are at most three maps with the same row and column la- bels, but different timestamps, in the bigtable. When a fourth is inserted, the one with the oldest label is dropped from the big table.

File: src/BigT/bigt.java [hunk 13]
```diff
@@ -0,0 +1,509 @@
...
+            btree.UnpinPageException, btree.FreePageException, btree.DeleteFileEntryException, btree.PinPageException,
+            PageUnpinnedException, InvalidFrameNumberException, HashEntryNotFoundException, ReplacerException {
+        String fileName = "";
+        heapFiles = new ArrayList<>(6);
+        indexFiles = new ArrayList<>(6);
+        heapFileNames = new ArrayList<>(6);
+        indexFileNames = new ArrayList<>(6);
+        this.name = name;
+        this.type = type;
+        boolean insert = (type ==1);
```
Comment (lowkya) [inline src/BigT/bigt.java:42]: type is an integer be- tween 1 and 5 and the different types will correspond to different clustering and indexing strategies you will use for the bigtable.

**Pull request merged**